### ISSUE TITLE
Reliable identification of unique client systems

### ISSUE THREAD

**Issue by Sebastian-Roth** [FOGProject/fogproject#198]
So far FOG used to identify clients via MAC address. Though MACs are pretty much unique all over the world we see more and more people using shared USB NICs to clone systems. This needs addressing.

First we added System UUID as an identifier but those [turned out](https://forums.fogproject.org/topic/10977/error-restoring-images-clients-having-identity-crisis) to be pretty [much unreliable](https://forums.fogproject.org/topic/10976/mac-address-problem-maybe).

**Comment by Sebastian-Roth** [comment]
Working towards a way to identify unique client systems reliably we need to combine the information we have. Though I don't think it's wise to add another field to the DB. Better if we just add the logic to the PHP code to use different identifiers - one after the other in case we still get too many results from the DB.

**Comment by Sebastian-Roth** [comment]
**Nevertheless IMHO this means that we have to give up on MAC address being a unique identifier in the DB** - which will be a major change. Therefore I started a new branch called [unique-identifier](https://github.com/FOGProject/fogproject/tree/unique-identifier) to work on this subject.

**Comment by astrugatch** [comment]
I am running into this exact issue. My work around has been to issue usb NICs to individual machines, but this is just that, a work around. If you need testers for this feature please let me know.

**Comment by Sebastian-Roth** [comment]
We had a long and intensive discussion on this topic and thought that essentially the hard disk is what makes a client a client from the FOG point of view. So I came up with some code to read HDD serial and WWN from the disk firmware. Adding that code to iPXE turned out to be a dead end as there is no disk driver/subsystem in iPXE yet and it would have been a huge task.

I then decided to go back to what we have, UUID, System Serial Number and Mainboard Serial Number. This should be good enough to uniquely identify most machines. Anyhow it's heaps better than what we have right now.

One big step is done. Added most of the code to identify hosts into the FOG PHP code and the FOS scripts. So this will be ready to test fairly soon.

**Comment by Sebastian-Roth** [comment]
People might be able to change their system UUID in some cases:
https://forums.mydigitallife.net/threads/bios-tools.529/page-19
https://kb.stonegroup.co.uk/how-to-change-the-dmi-information-a-stone-asus-based-system_357.html
https://kb.stonegroup.co.uk/how-to-change-the-dmi-information-on-a-stone-intel-nuc-system_412.html

**Comment by Sebastian-Roth** [comment]
We need to keep in mind that the MAC address is also used in other places in FOG, e.g. in `/images/dev/<MAC_WITHOUT_COLON>`. Not sure if we need to change that as well. There should not be a conflict as uploads are not as frequent and parallel as deploys are.

**Comment by astrugatch** [comment]
I have a shipment of machines that just came in with 3 spares so I can do some testing. I can spin up a new FOG instance with another branch if needed.

**Comment by danielbmiller** [comment]
A couple of things I have come across inventorying machines where I work (small public university, ~3000 units to date, hodgepodge of manufacturers and system configurations):

- Serial numbers are not universally unique
Prime examples of this are unset serial number fields (as discussed in the forums) but I have seen small companies that use literal serial numbers to identify their machines.

- Serial numbers are not necessarily unique within a company
I have run across companies that reuse/recycle serial numbers (Dell and IBM / Lenovo come to mind).  Instances have been few and far between and it has been quite a while since I have encountered one, but it is a thing.

Because of these issues, I have had to get things down to a model + serial number arrangement to (mostly) assure a unique identity.  For those that don't have programmatically accessible serial numbers I have resorted to manual entry, and for those that just don't have a serial number period, I have been known to create an arbitrary serial numbers and grab a sharpie.  Thinking about things, it would probably make sense to start using asset tags for the latter.

With this in mind along with the other issues brought up in the forums, it seems like using all the available information for the static components (what defines a host) might be useful. With this many values in play, programmatically choosing one is an option, or could take them all at once.  I'm not very familiar with iPXE, but if a cryptographic hash function is available to the init environment, running a string of those values through that would capture the uniqueness without needing to decide which identifier to use or having to keep track of multiple values. Wouldn't be without it's downsides (troubleshooting and reliably reproducing the identifier in the client come to mind), but could potentially result in more robust identifiers.

Just another spitball to consider.

**Comment by astrugatch** [comment]
I think danielbmiller is on the right track. A hash of a 2 or more readable fields may be the best move. This way you can get away with one being the same across models without getting the same identifier. MAC address should still NOT be one of the pieces used since this is the item that is no longer consistent on units that ship without ethernet ports.

**Comment by junkhacker** [comment]
this has been discussed and won't work as well as it looks like on the surface. i've seen plenty of systems were the only information that's available in the PXE environment, aside from MAC address, is identical (or blank) for all systems of that make and model. so the hashes would match multiple machines.

**Comment by astrugatch** [comment]
What information is available in the PXE environment? 

Do we have good examples of what other management/ imaging systems use as identifiers?

**Comment by danielbmiller** [comment]
More spitballing.... 
(and yes I am as much trying to get the information from posts spread out across the forums here as I am trying to spark discussion)

I think http://ipxe.org/cfg has a list of default attributes.
Looking through my last Ghost server (v2.5, some machines I have can't PXE), looks like they use mac addresses; I seem to remember Ghost getting quite irritated when I had machines with multiple NICs. 

If the data isn't accessible at the iPXE stage, can the non-unique identifier be classed as problematic to be handled as an edge case?  
For those identified edge cases, can the data available to iPXE be changed so that the uniqueness issue goes away, possibly via asset tag? I believe the devs behind iPXE say it can't be done permanently from within iPXE, but A) could FOS perform the changes (with smbios-utils or something similar,) B) is that something that FOS should do, and C) would this approach even work for the kinds of devices that have ambiguous identifiers? 

Could the TPM be leveraged to either locate, generate, or store an identifier? theoretically, there is a unique-ish RSA key that could be used to encrypt a known string...  But then again, are the problematic models likely to have a TPM and could that TPM be interacted with during a PXE session?

**Comment by astrugatch** [comment]
I suppose the bigger discussion needs to be about what is the acceptable edge case. Currently systems without built in Ethernet are considered edge cases by FOG, and that was fine pre ultra books and high speed wireless becoming the norm. My question would be if UUID could be made to work (I know @Sebastian-Roth mentioned issues with that code in this thread) are the manufacturers that don’t adhere to unique UUIDs the new acceptable edge case?  And a sub question would be who isn’t adhering to that? If say Dell, HP, Lenovo, Acer and Asus all follow a standard, but beige box or Aliexpress machines don’t I would personally see that as acceptable.

**Comment by Sebastian-Roth** [comment]
@astrugatch The most valuable information on this you can find in the forum: https://forums.fogproject.org/topic/10987/what-can-we-do-when-we-don-t-trust-uuid

**Comment by astrugatch** [comment]
@Sebastian-Roth 

I just looked over that post. I did not realize how all over the map manufacturers are. I come from a mostly Apple environment so standardized inventory isn't really an issue.

Based on what I see there Motherboard serial + HDD Serial seems like the most non-network base combination.

I know it was mentioned in that thread that HDD can be replaced / change, but the computer could be reenrolled at that point and added to the appropriate group (assuming persistent groups are used) and they are back in business. I feel like systematically thats easier than hunting down and making sure dongles remain paired to computers or constantly shifting mac addresses around during imaging.

**Comment by junkhacker** [comment]
HDD Serial is available within FOS, but not iPXE

**Comment by danielbmiller** [comment]
Looks like Tivoli Provisioning Manager uses a four part key consisting of IP address, MAC Address, UUID, and Serial Number with the requirement that at least one value be non-null.

Was googling through some things and it looks like smbios may be generally exposed in iPXE, just not with nice names all around.  see:
https://github.com/ipxe/ipxe/pull/21#issuecomment-45793439
http://etherboot.org/wiki/commandline

Of note is ${smbios/x.y.z} form for arbitrary access.  Assuming this is route is viable, the bear would be going through and determining what x.y.z is for the attributes in which we would be interested from the DMTF standard ( https://www.dmtf.org/sites/default/files/standards/documents/DSP0134_3.1.1.pdf ).

**Comment by Sebastian-Roth** [comment]
@ghooper90 said:
> So all of the bits needed to create the sysuuid value seem to be in there the just need to be glued together.

While you are mostly right on this (SMBIOS info available but md5/sha/base64 not available as simple string manipulation commands I think) I still feel that there is no consensus about which details to use.

I would consider the MAC address still being part of it as it is one very reliable detail on probably 90% of the machines. But @astrugatch argues against this for a good reason as well.

HDD serial is pretty much out I reckon! And I say this for a good reason. I have looked into adding HDD serial to the iPXE code. I started off by implementing it in C code to run on Linux and it worked great. Then I moved on to see how to access HDD information in the low level environment iPXE is running on early boot. I even talked to the iPXE devs but there is not much interest in this because they don't consider local disk access (of any form) valuable on PXE boot. I don't remember the details but I think NVMe/AHCI/RAID and such things would make it very hard to come up with reliable code anyway.

**Comment by danielbmiller** [comment]
Assuming there isn't a desire to collect information beyond that which FOG is currently storing for inventory purposes, unioning the attrubute sets from https://forums.fogproject.org/topic/10987/what-can-we-do-when-we-don-t-trust-uuid and excluding the sections that i would expect to be (relatively) easy to change or not natively accessible to PXE (Hard disk, Memory, Bios, and I seem to remember some discussion of CPU replacements as well) would leave the attributes related to System, Motherboard, and Chassis attributes.
> System Manufacturer	
> System Product	
> System Version	
> System UUID	
> System Serial Number	
> System Type	
> Motherboard Manufacturer	
> Motherboard Product Name	
> Motherboard Version	
> Motherboard Serial Number	
> Motherboard Asset Tag	
> Chassis Manufacturer	
> Chassis Version	
> Chassis Serial	
> Chassis Asset	

Of those attributes, I believe the Serial Numbers (system, mb, and chassis), UUID, and Asset Tags (mb and chassis) would successfully identify almost all of the machines I have deployed (some notable exceptions are those with replaced mobos that didn't have their SNs re-entered by whomever serviced the machine). It also appears that set of attributes would be sufficient for uniquely identifying most of the examples given in the UUID thread, but it is hard to tell with the redactions. Having the asset tags in the mix would provide an means to force some kind of uniqueness should all else fail.  I would like to see some sort of product identification in there as well, but with the lack of consistency of those identifiers, I don't know if I could convince myself that adding in 5 or more fields to that initial query would be worth guarding against a very small set of cases.

Another idea ... how accessible are the non-active mac addresses from iPXE? Possibly not very ( http://forum.ipxe.org/showthread.php?tid=6886&pid=9054 ) but maybe in future?

Other ideas?

Then I guess another issue to address is how to both identify and handle a machine isn't uniquely identified by the data available.

**Comment by mastacontrola** [comment]
My worry is that there's so many "options" but none of them seem to be standardized. MAC Address, in my opinion, is by far the most "stable" method when comparing what we're looking at.

Even HDD's sometimes don't have elements set appropriately.

That being said, I do agree that we need a better method of identifying hosts than using MAC Address with so many machines coming without NICs on board these days.

The best option I can think of is to use the HDD serial. Ultimately, the image is placed on the HDD. The HDD resides in the host. If you have to replace the HDD, You have to replace the image on the machine to begin with. Could we focus on just the HDD elements? If we need to, create a GUID generator based on the HDD's serial number.

Just spitting this out there.

**Comment by mastacontrola** [comment]
I am fairly sure VM's generate random serial numbers for the HDD as well. To be fair, i haven't looked at the fog inventory as we do try to get the inventory item for it, but I do remember seeing Serials.

To test: Zen, KVM, VirtualBox, HyperV. I Usually work with VMware.

**Comment by ghooper90** [comment]
I really think leaving the HD serial numbers out of the composite calculation is best. As Sebastian said nvme don't all seem to have serial numbers to reference. When I get the chance I'll sort out what fields would be best to use. As mentioned above I did want to include both the board and chassis asset tags to the FOG Admin can at the very least break the tie if 2 system's sysids are calculate the same.

**Comment by Sebastian-Roth** [comment]
> As Sebastian said nvme don't all seem to have serial numbers to reference.

Although this is not what I meant/said I still think that HD serial will cause us (mainly me, I suppose) hell of a lot of work that we/I better put into other things. One issue I see about NVMe is that accessing HD serial will work different than on classic HDs/SSDs. I don't even want to think about coming up with C code in iPXE to support all those different disk models (meaning SATA HD vs. PCIe NVMe!!)

As well HD serials in VMs is highly likely to not be unique. A quick search foo on the web revealed this:
https://github.com/rockstor/rockstor-doc/blob/master/quickstart.rst
> All drives must have unique serial numbers (real drives do); not all VM [*] systems default to this.

On the other hand question is: Are the other options set properly for VMs? I just checked for VirtualBox and `System UUID` seems fine. I just updated [George's initial post](https://forums.fogproject.org/topic/10987/what-can-we-do-when-we-don-t-trust-uuid). While for VMware UUID seems not available but `System Serial Number` seems unique.

**Comment by Sebastian-Roth** [comment]
@ghooper90 said:
> As mentioned above I did want to include both the board and chassis asset tags to the FOG Admin can at the very least break the tie if 2 system's sysids are calculate the same.

Definitely a good idea. I think we all agree to that!

**Comment by danielbmiller** [comment]
Would it also make sense to look into providing the capability to write asset tags for that purpose, either in FOS or some other solution?

**Comment by detay321** [comment]
I'm dealing with this issue right now. I'm assuming this feature has been integrated into the regular stable branch. If it has, how do I enable or change what the unique identifier?

**Comment by Sebastian-Roth** [comment]
@detay321 Finding a reliable unique system identifier turned out to be very hard and we never go to terms what to use and implement that.

You might want to explain what issue you are dealing with in more detail and most likely we can help you solve it another way.

**Comment by ahardylaniertech** [comment]
I've just now found this thread (lots of great info) and would love to assist in any way I can as this would be extraordinarily useful to my org.  Would it be an option to allow the admin to pick the unique system identifier options during the install?  In my case, all of my machines have a unique chassis serial number.  This would easily fix the issue for me.  However, at least one older model that I have does not have HDD serials (why?!?!).  Perhaps allow admins to pick the combined fields too?  I'm assuming this could all be handled within the php code at the time of registration (just combine vars) and within the client as well?  

Obviously this would be a temporary fix, but imaging PCs without physical nics is becoming SO common that this would be a nice workaround.  I'll dig into the php code and see what I can find.  We have PLENTY of test machines of varying models and makers.

**Comment by Sebastian-Roth** [comment]
@ahardylaniertech You are more than welcome to dig into this. I might give you a few pointers:
https://github.com/FOGProject/fos/search?q=uuid
https://github.com/FOGProject/fos/blob/master/Buildroot/board/FOG/FOS/rootfs_overlay/bin/fog.checkin#L96
...
https://github.com/FOGProject/fogproject/search?q=sysuuid
https://github.com/FOGProject/fogproject/blob/dev-branch/packages/web/lib/fog/fogbase.class.php#L526

Many years back when we tried to introduce this it turned out the sysuuid was not unique at all. Back then the code caused a lot of trouble for some people using the latest developer version and we jumped back but never got the act together to bring it back in and properly test on enough systems to make sure it does not fail us.

One very important point is that we can only use what is available in iPXE (as extending the iPXE code would be a huge effort on top and using a different network boot loader is even more scary to introduce now). Take a look at the [iPXE config menu](https://ipxe.org/cmd/config) on your machines to see what's available. To get into the iPXE config menu you can create a custom iPXE entry and call `config` from there.

**Comment by ahardylaniertech** [comment]
Awesome.  Thanks for the pointers.  I've got a steep learning curve in figuring out exactly how FOG code works, but I'll give it a shot.  I've pulled a copy of the code and likely will have several questions as I dig in.  Thanks for being so responsive already.

**Comment by ahardylaniertech** [comment]
Wow.  You really pointed me at the exact spot, ty!  What do you think about giving the admin the option to choose?  By default the system uses the MAC, but the admin could choose to change (at any point in time) to using a different method.  Only machines added from that point forward would be affected.  There could be a few options (MAC, UUID, chassis serial, a combination of those, or manual input during full registration.  It shouldn't require any changes to ipxe or any of that code, just the PHP side of things and perhaps the FOG client.  

I'm not clear on how the client identifies itself to the system . . I would assume MAC, right?, but does it submit a combo of vars like the ipxe client during registration (so we could handle on the php server side) or, I would expect more likely, a code change so it can determine the value it submits.

Interested in your feedback..  Ty!

**Comment by astrugatch** [comment]
I like the idea @ahardylaniertech presents with choosing. For us UUID or system serial is perfect because we stick with Dell and they adhere to those standards as far as I can tell. But YMMV with other vendors, but having the fall back on MAC is a solid option.

**Comment by Sebastian-Roth** [comment]
@ahardylaniertech said:
> What do you think about giving the admin the option to choose?

On the one hand side it would be good so people can adjust things to their needs. On the other hand it will make the coding more complex as well as prone to errors (e.g. unexpected stuff you can't know about beforehand). Also people might get the setting wrong causing trouble in their environment.

> By default the system uses the MAC, but the admin could choose to change (at any point in time) to using a different method.

I like the idea of leaving MAC as default to not effect all users straight away with such a fundamental change.

> Only machines added from that point forward would be affected.

I don't think this is going to work out. I wouldn't want to keep track of which method to use for each individual machine/host. Makes the logic even more complex.

> There could be a few options (MAC, UUID, chassis serial, a combination of those, or manual input during full registration)

While the first three seem fine from my point of view "manual input" is not going to work because we can't burn that information into a ROM in the machine/host for next time it boots up again. We need to use the information provided by each individual host.

> It shouldn't require any changes to ipxe or any of that code, just the PHP side of things and perhaps the FOG client.

Haven't spent too much time thinking this through yet. As I said, I wouldn't touch the iPXE code itself. PHP server code sure and also the FOS init scripts need adjustment (see links above). Not sure about the [fog-client code](https://github.com/FOGProject/fog-client).

> I'm not clear on how the client identifies itself to the system . . I would assume MAC, right?

The host/machine booting through iPXE sends a couple of vars/values (MAC, sysuuid, maybe more) but we only ever look at the MAC address on the server side (PHP code).

> but does it submit a combo of vars like the ipxe client during registration (so we could handle on the php server side) or

Not sure I understand this. If set to boot from network (PXE) a host would send "a combo of vars" (see my comment above) on each and every boot. Sure not as many vars/values as we send on registration (like disk size and so on) but definitely more than just the MAC. I think this is an important part, so please pay attention: We need to use what's available early in the very first stage of PXE booting because at that point we need to distinguish between hosts. We can't boot up to the point where the Linux kernel (bzImage) is loaded because decision making needs to happen before that (show menu, send to imaging task, ...).

To make a long story short: We need to identify hosts/machines at iPXE boot stage. Doesn't mean we need to dig into the iPXE code but we just have to use what's available at this stage of booting. Take a look at the [iPXE config command](https://ipxe.org/cmd/config) as well as [iPXE cfg vars/values](https://ipxe.org/cfg).

**Comment by AdamKearn** [comment]
Sorry to dig up an old thread but did you get anywhere with this?

I was thinking the UUID in IPXE (aka sysuuid in fog) would be best?

I've got a few USB network adapters as we have laptops without the network port.

Unless there is an easy way to exclude the MAC of the adapter so fog moves on and tries the WiFi mac instead for identification.

**Comment by darksidemilk** [comment]
@AdamKearn  The super duper easiest way, provided you're on the latest version of fog 1.5.x or 1.6, is to enable mac address pass-thru on the client's bios setting if it's available. Fog will see that internal mac instead of the usb adapter mac and the usb adapter never gets tied. We got this working in FOS (what fog boots to over the network) a little while back. This covers most scenarios.

Since not every device has that pass-thru setting available though, the next level method involves a little more work and automation. You can use the FogApi powershell module for adding and removing macs to hosts in command line. i.e. I have a json of my usb macs stored in an internal central location ( I use a proget asset repo). You can then build something with powershell to easily add your usb macs before imaging and you can add a task to your setupcomplete or after whatever provisioning method you have is done to remove the mac with the api. 

I'd start with the mac address pass thru settings though.

**Comment by AdamKearn** [comment]
@darksidemilk thanks for the quick reply.

I'll have a look to see if this option is available for us in the BIOS.

If not I was thinking about editing the full host registration script to override it.

I've already built a kernel that supports the WiFi adapter so I should be able to replace it using a bit of RegEx and the iw command in Linux if not.

**Comment by mastacontrola** [comment]
Picking this back up. Re-traced the current code against everything discussed here, and there's a cleaner path than where we left off in 2018.

### Key finding: most of what we wanted is already available at the iPXE stage

The thread's blocker was "we can't read serials at iPXE without writing C." That's only true for **HDD serial** (no disk subsystem in iPXE, and unreliable in VMs anyway — correct to drop it). But the SMBIOS System/Baseboard/Chassis fields are **native named settings in current iPXE**, no fork required:

| iPXE setting | SMBIOS source | inventory column |
|---|---|---|
| `${uuid}` | Type 1 System UUID | `sysuuid` |
| `${serial}` | Type 1 System Serial | `sysserial` |
| `${board-serial}` | **Type 2 Baseboard Serial** | `mbserial` |
| `${asset}` | Type 3 Chassis Asset Tag | `caseasset` |

`${board-serial}` is the one that matters most — motherboard serial was our most-reliable field, and iPXE has exposed it as a named setting all along. The boot menu currently only sends `param sysuuid ${uuid}`; it could send the rest with no iPXE changes.

### The server side is mostly already built

- `HostManager::getHostByUuidAndSerial()` already does weighted composite scoring with per-field invalid-value blocklists. It's just **never called** — the call site in `FOGBase::getHostItem()` is commented out (disabled years ago after duplicate-UUID grief on MSI boards).
- `getHostItem()` is the **single chokepoint** for host resolution across iPXE boot, FOS, the web UI, and the C# client.
- In the deploy path, once a host is resolved at boot, the server hands FOS the host's **stored primary MAC** on the kernel line (`bootmenu`, `self::$Host->get('mac')`), not the MAC the machine booted with. So if composite matching resolves the right host despite a shared USB-dongle MAC, FOS rides on the correct MAC downstream — **FOS likely needs no changes**, and the C# client can stay on MAC for now (the collision is an imaging-time problem, not steady-state).

### Why the original attempt was unsafe

UUID-first with no guard meant a duplicate/blank value could silently hijack a host's identity — strictly worse than the MAC problem. Re-enabling needs: unique-match-or-fall-back-to-MAC, a minimum score, and never identifying on `product`/`manufacturer` alone.

### Normalization (the silent-failure trap)

Inventory values are stored verbatim; iPXE-supplied values must compare byte-for-byte. Findings:
- **UUID byte-order**: non-issue on modern iPXE+dmidecode — both byte-swap for SMBIOS >=2.6 and converge on the same canonical string.
- **Case**: absorbed by the DB's case-insensitive collation.
- **Whitespace/placeholder variance**: the real risk, and cheap to neutralize.

### What I've done (branch `feature-198-unique-host-identifier`, off `working-1.6`)

Committed the normalization groundwork only — `working-1.6` is intentionally untouched:
- `FOGBase::canonicalizeIdentifier()` — one helper (trim + collapse whitespace), used on both write and match so they can't drift.
- `inventory.php` — canonicalize `sysuuid`/`sysserial`/`mbserial` on write.
- `getHostByUuidAndSerial()` — canonicalize incoming values, case-insensitive blocklists, canonical+case-folded scoring.

Host-resolution behavior is unchanged: the composite call site stays commented and the boot params still send only `sysuuid`. This branch is for **testing whether the SMBIOS approach suffices** before flipping it on.

### What's still needed (bench, not code)

1. Confirm on real hardware that `${board-serial}`/`${serial}`/`${asset}` resolve in the iPXE `config` screen across a few vendors (Dell/HP/Lenovo + a mini-PC + a VM).
2. Confirm those values match the stored `inventory` row byte-for-byte (`SELECT iSystemUUID, iSysserial, iMbserial, iCaseasset FROM inventory WHERE hostID=<id>`).

If that holds, the remaining work is: add the boot params, and re-enable the composite lookup with the unique-or-fall-back-to-MAC guard.

**Comment by darksidemilk** [comment]
In my experience the board serial and serial are different between versions, even getting the UUID from bios differed slightly between lenovo, dell, asus, and HP (universal eh?).
I did a quick grab of a couple different hosts in my FOG inventory and it looks like board-serial had a unique value on a 

- custom built machine with an asus board
- an HP laptop
- a lenovo laptop
- a microsoft surface go
- a hp mini desktop
- dell detachable tablet - had the sysserial in the mbserial with slashes i.e. `/sysserial/uniquembserial/`

There was no value or a value of 'none' on

- vmware vm - 'none'
- qemu/kvm vm - null/empty

Both the VMs did have a uuid filled though.

granted in some cases I have the inventory pulled and written to windows via the api and my fogapi function https://github.com/darksidemilk/FogApi/blob/master/FogApi/Public/Get-FogInventory.ps1 and https://github.com/darksidemilk/FogApi/blob/master/FogApi/Public/Set-FogInventory.ps1

Granted the cim/wmi paths should be yielding the same values, I recall making sure of that back when I wrote that.

**Comment by AdamKearn** [comment]
@mastacontrola I would also advise against the motherboards serial number as I've seen a few brands set generic serial numbers such as 000000000 or with custom builds "To Be Set By OEM"

I personally think the UUID that's generated by iPXE is preferable as this is unique and also not network adapter specific (like MAC address are)

I've been building my own version of FOS locally for testing and also learning.

I've included Realtek WiFi Drivers into my image so after booting it will connect to our Wireless network using a hidden SSID and start partclone as normal etc.

Using UUID will return the same "asset" so it will always match the same machine no matter the boot source LAN/WiFi/USB-Adapter.

**Comment by mastacontrola** [comment]
@darksidemilk @AdamKearn — thank you both, this is exactly the kind of real-world data this needs, and it actually clarifies the direction. Let me be explicit about the approach, because I don't think we're picking a single magic field.

### We are not chasing "one identifier that works"

Two things we're explicitly *not* doing:

1. **Not single-field.** Your two comments make the case for us. @AdamKearn favors UUID — but UUID came back blank on Surface Pro 4 and `FFFF…` on some MSI boards earlier in this thread. @darksidemilk's data shows `board-serial` is `'none'`/empty on both his VMs — yet **those same VMs had UUID populated**. Where one field fails, another tends to be present. No single field survives contact with a mixed fleet.
2. **Not a hash either.** Concatenating fields and hashing them (proposed earlier) is brittle for the reason @junkhacker called out years ago: one blank or duplicated field poisons the whole token, and you can't see which field disagreed. We're not doing that.

### What we *are* doing: weigh several independent signals, resolve to one host

We score the host against multiple identifiers kept **separate**, then require a unique winner:

| Signal | iPXE source | Notes |
|---|---|---|
| System UUID | `${uuid}` | covers VMs and many physical machines |
| Baseboard serial | `${board-serial}` | strong on physical, often blank on VMs |
| System serial | `${serial}` | vendor-dependent |
| Chassis asset tag | `${asset}` | **admin-settable** — the manual lever for the unfixable cases |
| MAC | `${net0/mac}` | fallback, *not* a scored field — it's the untrustworthy signal in the USB-dongle case |

Because the signals stay independent:
- A blank/garbage field (`000000000`, "To Be Set By OEM", `FFFF…`) is dropped by a per-field blocklist and simply contributes nothing — the others still discriminate. @AdamKearn, that's how your generic-mobo-serial concern is handled without abandoning the field.
- We only accept a **unique** top match; on a tie or low confidence we fall back to MAC rather than guess. That guard is what was missing in the 2018 attempt (UUID-first, no guard → silent host hijack).
- When *every* auto-populated field collides (cheap beige-box / unset-everything), the chassis asset tag is the one field a human can write. The unfixable-by-firmware case becomes "set an asset tag," not "FOG can't tell these apart."

So a machine is identified by the **combination** of what it reliably reports, not by betting on any one value — and it degrades gracefully field by field instead of all-or-nothing.

### One question that decides how hard normalization is

@darksidemilk — you noted the UUID "differed slightly between lenovo, dell, asus, and HP." That's the crux for matching, so I want to pin it down: when you saw that difference, were you comparing the **iPXE/dmidecode** value against FOG's stored inventory, or the **WMI/CIM** value (your FogApi path) against it? WMI, dmidecode, and iPXE can each render the UUID with different byte-order/case. What matters for boot-time matching is whether **iPXE and dmidecode agree** on the same box. If you can drop into the iPXE `config` screen on a Lenovo/Dell and compare `${uuid}`/`${board-serial}` to that host's `inventory` row, that single data point tells us whether normalization is trivial or a real per-vendor problem.

(Groundwork for the canonicalization is committed on branch `feature-198-unique-host-identifier`; `working-1.6` behavior is unchanged until this is proven out.)

**Comment by AdamKearn** [comment]
Ahh   I understand how you are planning on approaching this now.

I'll have a look at our different systems and manufacturers too and see if I can see any difference across brands too.

We have an about 5-6 thousand machines so I'll compare these across our network and let you know.

**Comment by darksidemilk** [comment]
I can't find an example now of where uuid would be null or it would be something like '12345678-9012-3456-7890-abcdefabcdef' 
My get-foginventory code actually has a bug in it where I'm still trying to grab the qualifiers property when uuid is invalid, qualifiers is part of wmi but it doesn't show up in ciminstance, but that's where I was finding other properties from the win32_computersystemproduct class that match UUID

```
 $compSys = Get-CimInstance -ClassName Win32_ComputerSystemProduct;
#...
 if ($compSys.UUID -notmatch "12345678-9012-3456-7890-abcdefabcdef" ) {
                $hostObj.inventory.sysuuid       = $compSys.UUID;
            } else {
                $hostObj.inventory.sysuuid       = ($compSys.Qualifiers | Where-Object Name -match 'UUID' | Select-Object -ExpandProperty Value);
            }
```

This used to use wmi before I updated it to ciminstance since wmi isn't compatible with pwsh 7 and I missed this little tidbit.
I thought it was lenovo that did this and had a separate property where it was defined correctly, but the one I just checked this was not the case. I'll try to find which vendor this was for when I have some more time. It is possible that the computers that did this are gone from my environment and newer ones don't have this problem anymore, which just means we might have to get more input from people with older devices.

**Comment by AdamKearn** [comment]
I've just looked at our inventory table on the database and it seems that the serial number is quite inconsistent on custom Asus motherboards and older HP desktops. (2008-2010s)

We do have a model of laptops called "Stone Books" these seem to have a weird UUID with them all being prefixed with a specific string but the last part is random.

For example I'm seeing a few hundred devices prefixed with f090000-0000-1000-8000-XXXXXXXX

It only seems to be this model that we have currently.

All other models are showing a fully unique UUID without any prefixes etc.

I'm assuming this is part of the way iPXE is generating this.

Some of the other columns in this table are showing empty values or generic strings like "Enter Serial"

The UUID is always there without any duplications or NULL values in my environment.

Hope this helps - let me know if you need any more examples of "real world data"

**Comment by mastacontrola** [comment]
## Status, 2026-09-02: firmware identity is in working-1.6, in log mode

Landed by #1668, #1669 and #1672 (schema 413, live at 1.6.0-beta.5053 in the lab):

- **Every PXE boot sends four SMBIOS fields** alongside the MAC list: system UUID, system serial, baseboard serial and chassis asset tag, from iPXE's native `${uuid}`, `${serial}`, `${board-serial}`, `${asset}`. No iPXE change was needed.
- **A pure scoring rule decides** (`SmbiosIdentity::pick()`): one point per matching field, compared per field on canonicalized values; empty, known-placeholder and single-repeated-character values score nothing; the winner must hold the top score alone, a tie is "no opinion"; the asset tag breaks ties but cannot win by itself. This is what the 2018 attempt lacked (the MSI `FFFFFFFF-…` case).
- **`FOG_HOST_IDENTIFY_SMBIOS`** is a three-way setting shipped as **`log`**: the MAC still decides and every disagreement is written to the web error log prefixed `FOG host identity`. `enforce` lets a unique firmware match win. `off` ignores the fields.
- **Boot fills the inventory row.** A host created by CSV, API or the UI has no inventory row until its first full inventory; the boot path now creates it with the identity fields, and fills empty ones on an existing row. It never overwrites a value FOS stored; a disagreement between iPXE and dmidecode is logged instead.
- The four inventory columns are indexed; the lookup runs on every boot request.

Lab proof: an EFI VirtualBox VM with crafted DMI values (Dell-style `/serial/…/` board serial, all-digit asset tag) PXE-booted and its inventory row held all three values byte-for-byte within 15 seconds.

The decision runs on the iPXE boot path only. FOS inherits it, because the boot menu hands FOS the resolved host's stored primary MAC on the kernel line. The FOG client, the API, CSV import and registration still key on MAC.

### Still open on this issue

1. Collect `FOG host identity` log lines from real fleets (the placeholder survey above is what they will show) and decide when the default can move from `log` to `enforce`.
2. Extend the same decision to the FOG client's check-in, which means the client reading SMBIOS and sending it, and the server applying `SmbiosIdentity` under the same setting.

Design record: `docs/adr/0039-a-booting-machine-is-identified-by-firmware-second.md` on `working-1.6`.

Branch cleanup: `feature-198-unique-host-identifier` (2018-era attempt, last tip `145729d73`) has been deleted; its content was re-applied and superseded by #1668.

**Comment by mastacontrola** [comment]
Registration now asks the firmware too: #1674 (server, merged into working-1.6) and FOGProject/fos#171 (FOS sends all four fields with a registration, merged). Under `enforce`, a machine whose MAC is unknown but whose firmware names exactly one host is attached to that host instead of registering as a duplicate; a known MAC still wins in every mode. Proven against the lab database (10/10, see the PR). Open item 2 above is now the FOG client's check-in only; the MAC-less host model is recorded in ADR 0039 as its own decision.

**Comment by mastacontrola** [comment]
## Proven: the firmware identity does not depend on a MAC at all

A question came up about how far the new identity actually goes: if a host has
**no MAC addresses on it whatsoever**, does the firmware still find it? Tested
on a lab server running `1.6.0-beta.5257`, and the answer is yes -- it works
exactly as ADR 0039 designs it.

The setup: host `claudeid1` created normally, PXE-booted once so the boot path
seeded its `inventory` row (`sysuuid=5c1a17de-0000-4000-8000-000000000198`,
`sysserial=CLAUDEID1SER`), then **every** `hostMAC` row for it removed. The host
still loads and is valid with zero MACs -- `FOGController::buildQuery()` builds
the `MACAddressAssociation` relation as a `LEFT OUTER JOIN` with the
`hmPrimary='1'` filter in the `ON` clause rather than the `WHERE`, so no primary
MAC means a null column, not a missing row.

Booting it again from a MAC the server has never seen (`02:00:00:ca:1d:99`):

**`log` mode** -- the MAC decides, so the machine gets the unregistered menu,
and the disagreement is reported:

```
FOG host identity (log): MAC resolved no host; SMBIOS resolved "claudeid1" (id 264) on sysuuid=5c1a17de-0000-4000-8000-000000000198, sysserial=CLAUDEID1SER; MAC kept (log mode)
```

**`enforce` mode** -- same boot, and the menu comes back
`Host is registered as claudeid1!`:

```
FOG host identity (enforce): MAC resolved no host; SMBIOS resolved "claudeid1" (id 264) on sysuuid=5c1a17de-0000-4000-8000-000000000198, sysserial=CLAUDEID1SER; SMBIOS wins
```

That is the whole point of the design confirmed end to end:
`HostManager::resolveHostBySmbios()` queries `inventory` and hands the rows to
`SmbiosIdentity::pick()`, and neither one reads a MAC anywhere. The MAC is the
default identity, not a prerequisite for the firmware one.

### Two things this is not

**It is not a supported state to be in.** You cannot reach zero MACs through
the UI or the API -- `hostMacaddressPost()` refuses with `Cannot delete the
primary mac address, please reselect`, and on a single-MAC host the primary is
the only MAC, so there is nothing to reselect. The state above had to be forced
with a direct model call. It is a proof of where the identity logic keys off,
not a configuration to run.

**A MAC-less host is found but cannot be tasked.** Every task hands FOS the
host's *stored* primary MAC on the kernel line, so a host with none gets
`mac=` empty. This does not affect the case the feature exists for: a machine
whose NIC changed still has its old primary MAC on file, so `mac=` is populated
and FOS's check-in resolves back to the same host. Only the artificial
zero-MAC host hits it.

Still ships as `log`. Nothing here argues for moving that default -- that still
waits on `FOG host identity` lines from real fleets.
